File: src/api/member.ts
import { invoke } from "@tauri-apps/api/core";
import type {
  CreateMemberPayload,
  MemberDetail,
  MemberItem,
  MemberListQuery,
  MemberListResult,
  UpdateMemberDetailPayload,
  UpdateMemberPayload,
} from "@/types/member";

export async function pingMember(): Promise<string> {
  return invoke<string>("member_ping");
}

export async function memberCount(): Promise<number> {
  return invoke<number>("member_count");
}

export async function listMembers(query: MemberListQuery): Promise<MemberListResult> {
  return invoke<MemberListResult>("list_members", { query });
}

export async function listMemberGroups(): Promise<string[]> {
  return invoke<string[]>("list_member_groups");
}

export async function createMember(payload: CreateMemberPayload): Promise<MemberItem> {
  return invoke<MemberItem>("create_member", { payload });
}

export async function updateMember(payload: UpdateMemberPayload): Promise<MemberItem> {
  return invoke<MemberItem>("update_member", { payload });
}

export async function deleteMember(id: number): Promise<void> {
  return invoke<void>("delete_member", { id });
}

export async function getMemberDetail(id: number): Promise<MemberDetail> {
  return invoke<MemberDetail>("get_member_detail", { id });
}

export async function updateMemberDetail(
  payload: UpdateMemberDetailPayload,
): Promise<MemberItem> {
  return invoke<MemberItem>("update_member", { payload });
}

File: src-tauri/src/commands/member.rs
use crate::data_access::member;
use crate::services;
use crate::state::AppState;

#[tauri::command]
pub fn member_ping() -> &'static str {
    "member"
}

#[tauri::command]
pub async fn member_count(state: tauri::State<'_, AppState>) -> Result<u64, String> {
    member::member_count(&state.db, &state.cache)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_members(
    state: tauri::State<'_, AppState>,
    query: services::member::MemberListQuery,
) -> Result<services::member::MemberListResult, String> {
    member::list_members(&state.db, &state.cache, query)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_member_groups(state: tauri::State<'_, AppState>) -> Result<Vec<String>, String> {
    member::list_member_groups(&state.db, &state.cache)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn create_member(
    state: tauri::State<'_, AppState>,
    payload: services::member::CreateMemberPayload,
) -> Result<services::member::MemberDto, String> {
    member::create_member(&state.db, &state.cache, payload)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn update_member(
    state: tauri::State<'_, AppState>,
    payload: services::member::UpdateMemberPayload,
) -> Result<services::member::MemberDto, String> {
    member::update_member(&state.db, &state.cache, payload)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn delete_member(state: tauri::State<'_, AppState>, id: i32) -> Result<(), String> {
    member::delete_member(&state.db, &state.cache, id)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_member_detail(
    state: tauri::State<'_, AppState>,
    id: i32,
) -> Result<services::member::MemberDto, String> {
    member::get_member_detail(&state.db, &state.cache, id)
        .await
        .map_err(|e| e.to_string())
}
